File: exercise/215/quick.cpp
class Solution {
public:
    void swap(vector<int> &nums, int a, int b) {
        int t = nums[a];
        nums[a] = nums[b];
        nums[b] = t;
    }

    int quick(vector<int> &nums, int k, int begin, int end) {
        if (begin == end - 1)
            return nums[begin];
        int q = nums[end-1];
        int left = begin;
        for (int i = begin; i < end - 1; i++) {
            if (nums[i] > q) {
                swap(nums, i, left);
                left++;
            }
        }
        swap(nums, end - 1, left);
        if (left == k) {
            return nums[left];
        } else if (left < k) {
            return quick(nums, k, left + 1, end);
        } else {
            return quick(nums, k, 0, left);
        }
    }

    int findKthLargest(vector<int>& nums, int k) {
        return quick(nums, k - 1, 0, nums.size());
    }
};

File: exercise/215/quick.py
class Solution:
    def quick(self, nums: List[int], k: int, begin: int, end: int) -> int:
        mid = nums[end]
        left = begin
        for i in range(begin, end):
            if nums[i] > end:
                nums[i], nums[left] = nums[left], nums[i]
                left += 1
        nums[end], nums[left] = nums[left], nums[mid]
        if left == k:
            return nums[left]
        elif left < k:
            quick(nums, k, left + 1, end)
        else:
            quick(nums, k , begin, left - 1)

    def findKthLargest(self, nums: List[int], k: int) -> int:
        return quick(nums, k - 1, 0, len(nums) - 1)

